Modernize Elasticsearch mappings (part 1)

joxl opened the pull request:
## Technical Summary

Begin bringing our Elastic mappings (in code) up to speed with how they actually exist in Elasticsearch.  This change converts all the deprecated `"type": "multi_field"` property mappings to their modern equivalents in preparation for the Elastic v5.x upgrade (v5.x strictly forbids the `multi_field` type).

Jira: [SAAS-12152](https://dimagi-dev.atlassian.net/browse/SAAS-12152)

## Safety Assurance

### Safety story

New tests have been added. Many existing Elastic-related tests are now labeled `es_test` (where applicable).  All linted mappings were tested for equality against their previous unlinted version.  The new "remove 'multi_field' types" mappings have been checked for accuracy against real, live Elastic indices.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

[QA-4040](https://dimagi-dev.atlassian.net/browse/QA-4040)

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

The included migration will update all mappings to the new syntax.  Before/after observations will be taken on both staging and prod to ensure no unexpected mapping changes will be performed.

The mapping updates do not require a reindex, and if a reindex were needed with these new mappings, it would be safe to do so because there are no "functional changes" in these mapping updates.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

joxl committed:
Lint pillows.core

joxl committed:
Convert JSON mappings to python

- app
- case_search

joxl committed:
Mark another flaky test.

joxl committed:
Add management command for printing Elasticsearch index mappings.

joxl committed:
Lint and sort mappings in preparation for changes

All mappings are equal (identical) to previous versions, only whitespace, order
and syntactical values have changed.

joxl committed:
Use global date formats for domain, group, and report_case mappings.

All mappings are equal (identical) to previous version. The date formats
were already equal to the global values, but were not linked.

joxl committed:
Add transforms to mapping printing utility

joxl committed:
Update get_all_expected_es_indices() to use authoritative index collection

joxl committed:
Add `es_test` decorator to unlabeled Elasticsearch tests

joxl committed:
Update unique `case_search` mapping

The "default" `domain` property is not indexed, which is different from
all other (current) uses of the `multi_field` datatype.

joxl committed:
Update Elastic mapping `multi_field` property types

- Convert deprecated `multi_field` properties to the modern equivalent.

joxl committed:
Add ability to print mappings fetched from Elasticsearch index.

This is a portion of an attempt at writing tests to verify that the
Elastic mappings in code match the ones in the actual Elastic indices.
That attempt failed miserably because Elasticsearch only seems to return
mapping properties for _fields present on indexed documents_ rather than
the _full mapping the index was created/updated with_.

snopoke commented on corehq/pillows/management/commands/utils.py:
```diff
@@ -0,0 +1,126 @@
+        stream.write(chunk)
+
+
+class ElasticMappingEncoder(json.JSONEncoder):
```
docstring explaining why this is needed would be great

joxl replied:
Docstrings added (new commit).

snopoke reviewed (commented):
Looks OK. What's the purpose of the new JSON serializer? 

Outside of manual verification of the mapping changes would it not be worth doing some kind of testing on staging?

millerdev reviewed (commented):
Nice! Lots of good cleanup in here. The only important comment is the one about `finally`. All of the other nits can be ignored at your discretion. Other wise looks good to me.

:lotus_position: :world_map:

joxl committed:
Add docstrings for ElasticMappingEncoder and related utils.

joxl commented on corehq/pillows/mappings/tests/test_utils.py:
```diff
@@ -0,0 +1,22 @@
+    def test_sorted_mapping(self):
+        key_order = ["alpha", "items", "zulu", "properties"]
+        mapping = sorted_mapping(self.mapping)
+        self.assertEqual(key_order, list(mapping.keys()))
+        self.assertEqual(key_order, list(mapping["items"][0].keys()))
+        self.assertEqual(key_order, list(mapping["properties"].keys()))
```
:man_facepalming:

joxl committed:
Apply suggestions from code review

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

joxl commented:
> Outside of manual verification of the mapping changes would it not be worth doing some kind of testing on staging?

That's a good idea. I'm not sure how involved it is to do a reindex.  I'll look into this.

calellowitz commented on corehq/pillows/management/commands/utils.py:
```diff
@@ -5,16 +5,44 @@
+    """Encodes an Elastic mapping (nested python dict) as properly indented
+    python code.
 
+    A JSONEncoder which writes 'True', 'False' and 'None' instead of 'true',
```
This is probably a naive question, but the description here sounds a lot like just calling `pprint.pp` (or one of the other pretty print functions) on the dict object, instead of `json.dump`. What are differences between this implementation and that?

joxl replied:
I investigated both the `pprint` module, as well as `IPython.lib.pretty` (used by Django shell) prior to this implementation. Neither of those modules produce this familiar, readable format. Also, neither are extensible (would have required re-writing their object traversal logic to change the whitespace, which defeats the purpose).

calellowitz reviewed (commented):
Besides one question, the mapping changes themselves look good. Had a few stylistic nits, as always those are up to you, and a few questions for my own understanding on the new tooling, which I am still trying to wrap my head around.

joxl commented on corehq/pillows/management/commands/print_elastic_mappings.py:
```diff
@@ -0,0 +1,47 @@
+
+    def add_arguments(self, parser):
+        parser.add_argument("-o", "--outfile", metavar="FILE",
+            help="write output to %(metavar)s rather than STDOUT")
```
The `argparse` module is what's performing the format operation on this string, not this code.  I don't know of any way to use f strings for this.

calellowitz replied:
Yep, sorry. Don't know how I missed that.

joxl commented on corehq/pillows/management/commands/print_elastic_mappings.py:
```diff
@@ -0,0 +1,47 @@
+        else:
+            namespace = MAPPING_SPECIAL_VALUES
+        if outpath is None:
+            outpath = self.stdout._out  # Why, OutputWrapper.write()? Why?
```
The `self.stdout` (an instance of `OutputWrapper`) doesn't comply with Duck-typing because its `write()` method automatically writes a trailing newline every time it's called (behaves like `print()`), so doing things like passing it to a JSON encoder has awful results.

snopoke replied:
Does settings `self.stdout.ending = None` not solve the newline problem?

joxl replied:
Hmm, that would fix the line-ending problem (I didn't know about this `ending` property). I'll also have to contend with the `style_func` function.

I decided to `sys.stdout` instead of `self.stdout`.  This gains me two things:

1. Allows use of `argparse.FileType` for the `--outfile` argument which simplifies the code as @calellowitz recommended.
2. Don't have to muck with Django's `OutputWrapper.ending` and `OutputWrapper.style_func` for guaranteed output consistency.

The [Django docs](https://docs.djangoproject.com/en/2.2/howto/custom-management-commands/) mention that the importance of using `self.stdout` over `sys.stdout` is for testing the output, which is not a concern (`sys.stdout` is simple enough to patch in tests).  Note that `sys.stdout` is necessary with the `FileType` argument type because it explicitly uses `sys.stdout` if the special value `-` is provided as the argument value.

joxl commented on corehq/pillows/mappings/case_mapping.py:
```diff
@@ -206,13 +194,9 @@
                     "index": "not_analyzed",
                     "type": "string"
-                },
-                "type": {
```
The "deleted" fields [are the "top level" property](https://www.elastic.co/guide/en/elasticsearch/reference/1.3/_multi_fields.html). I say "deleted" because they're not really deleted.  The old `multi_field` type had a [special behavior](https://www.elastic.co/guide/en/elasticsearch/reference/0.90/mapping-multi-field-type.html#_accessing_fields) that the field _named the same as the property_ (the "default" field) was the one accessible without its `.<fieldname>` suffix.  Elasticsearch versions 1.x and 2.x have always silently upgraded mappings to the current format (for example, see: http://localhost:9200/smslogs/_mapping/field/domain), but version 5.x and up strictly prohibit the `multi_field` type.

calellowitz replied:
Thanks! That makes a lot of sense

joxl commented on corehq/pillows/core.py:
```diff
@@ -1,22 +1,17 @@
 
-#https://github.com/elasticsearch/elasticsearch/issues/2132
-#elasticsearch Illegal pattern component: t
-#no builtin types for || joins
```
Correct, no longer relevant: https://github.com/elastic/elasticsearch/issues/2414

joxl commented on corehq/pillows/management/commands/print_elastic_mappings.py:
```diff
@@ -0,0 +1,47 @@
+            help="print mapping for %(metavar)s")
+
+    def handle(self, cname, **options):
+        outpath = options["outfile"]
```
There are merits to having an argument-specified output file as apposed to depending on your shell to do it, which is a feature I was specifically aiming for here.  The added complexity isn't many lines of code, and this command is quite short even with that added complexity.  Are there specific parts that are hard to follow? Maybe I could improve it by adding some comments?

joxl committed:
Use `sys.stdout` as default command output file instead of `self.outfile`

- allows using `argparse.FileType` to simplify file handling
- eliminates hacking around `OutputWrapper`'s lack of easy duck-typing

joxl committed:
Lint

calellowitz reviewed (approved):
Thanks for the answers, and the changes to the command. This looks great! Approving now because all of it looks fine, but I think I am missing some backstory on the printing command. You mention in a few places some very specific requirements, but it is not totally clear to me what those requirements. I would love to hear a bit more about how all this will be used, but definitely fine to merge in the mean time since the code itself looks good. I will agree with Simon's comment about testing on staging though. These kinds of cross cutting changes, even those that are mechanical and seem clearly safe like this, are nice to sanity check in a clustered environment, rather than just docker.

joxl commented:
Thanks for the feedback @calellowitz.

> I think I am missing some backstory on the printing command.

The printing command has a few functions:

1. Printing the mappings in a consistent format.
2. Using the verbatim output to update mapping source code.
3. Generating transformed mappings (e.g. for upgrading mappings to modern formats) with an iterative workflow.

This would probably be valuable info for the description or epilogue of the command itself. I'll add it there also.

> You mention in a few places some very specific requirements, but it is not totally clear to me what those requirements.

This is a general practice I follow when writing any commandline utility that is designed to output meaningful content beyond just a human looking at it. For such tools, it is common and convenient for the utility to provide the capability to write "rendered" output to an argument-specified file.  I could elaborate further on the shell-specific justifications for this, but I'll refrain from those for now. Another benefit is that an explicit output option gives the user confidence that specifying a named file as an argument will result in only _data_ being written in that file (and not other logging output).  I've used HQ management commands designed to output CSV data, which do not provide an option for writing that data to a named file, and also write logging text to STDOUT.  As well as being useful, having an output file argument present is a helpful indicator to the end user that "pure output" is a feature of this utility.

joxl committed:
Add documentation on the purpose of the `print_elastic_mappings` command

joxl committed:
Merge branch 'master' into jm/elastic-modernize-mappings

joxl committed:
Add migration to update mappings

joxl committed:
Prune `domain` mapping as per #31207 (missed in conflict resolution)

joxl committed:
Update `update_es_mapping` command with new options

- `--dry-run`: perform a dry-run (useful for observing expected change
  diff)
- `--no-diff`: disable diff output
- `--quiet`: disable all non-error output

joxl commented:
Commit 1fdd923d51f should reference #31211, the PR in the commit message is the one that actually reindexed for the mapping change.

joxl committed:
Don't prompt on --dry-run

millerdev commented on corehq/pillows/migrations/0001_elastic_v5_mapping_compatibility.py:
```diff
@@ -0,0 +1,12 @@
+    dependencies = []
+
+    operations = [
+        update_mapping_migration_op(info) for info in get_all_expected_es_indices()
```
I think it would be better to hard-code the values of `CANONICAL_NAME_INFO_MAP` here rather than using a utility function since it may not be obvious to future maintainers how changing that function or the contents of `CANONICAL_NAME_INFO_MAP` will affect this migration. For example, if a new item is added to `CANONICAL_NAME_INFO_MAP` long after this migration is created, should this migration process that item? Probably not, and a new migration will be needed for where this migration has already been applied anyway.

This means adding lines like the following to this migration:
```py
from ..mappings.app_mapping import APP_INDEX_INFO
from ..mappings.case_mapping import CASE_INDEX_INFO
```
and then changing this to something like
```py
    operations = [
        update_mapping_migration_op(APP_INDEX_INFO),
        update_mapping_migration_op(CASE_INDEX_INFO),
        ...
```

joxl replied:
Yeah, this totally dodged my radar. Thanks for pointing it out.

millerdev commented on corehq/pillows/utils.py:
```diff
@@ -89,3 +93,17 @@ def get_all_expected_es_indices():
     return form_metadata
+
+
+def update_mapping_migration_op(index_info, quiet=True):
```
What do you think of changing the name of this argument to `index` and passing in `index_info.index`?

joxl replied:
Good catch, previous (squashed) versions of this needed the info object. I missed this simplification after cleaning it up.

joxl committed:
Move `fetch_elastic_mapping()` out of tests/utils (used by manage cmd)

joxl committed:
Code review: `update_mapping_migration_op()` doesn't need index_info obj

joxl committed:
Code review: be explicit about what needs to be migrated

joxl commented:
Split out into #31424 and #31425. Please resume reviews there.
